File: src/main/cpp/subsystems/Spindexer.cpp
#include "subsystems/Spindexer.h"

#pragma region Spindexer
/// @brief Constructor for the Spindexer subsystem
Spindexer::Spindexer()
{
    // Configure the spindexer motor
    TalonFXConfiguration(&m_spinnerMotor,
                         20.0_A,          // Maximum Amperage
                         true,            // Inverted
                         false,           // Brake mode
                         false,           // Continuous wrap
                         0.4,             // P
                         0.0,             // I
                         0.0,             // D
                         0.0,             // S
                         0.0,             // V
                         0.0,             // A
                         0_tps,           // Velocity Limit
                         0_tr_per_s_sq);  // Acceleration Limit

    // Configure the kicker motor
    TalonFXConfiguration(&m_kickerMotor,
                         20.0_A,          // Maximum Amperage
                         false,           // Inverted
                         false,           // Brake mode
                         false,           // Continuous wrap
                         0.1,             // P
                         0.0,             // I
                         0.0,             // D
                         0.0,             // S
                         0.0,             // V
                         0.0,             // A
                         0_tps,           // Velocity Limit
                         0_tr_per_s_sq);  // Acceleration Limit

    TalonFXConfiguration(&m_kickerFollowerMotor,
                         20.0_A,          // Maximum Amperage
                         false,           // Inverted
                         false,            // Brake mode
                         false,           // Continuous wrap
                         0.1,             // P
                         0.0,             // I
                         0.0,             // D
                         0.0,             // S
                         0.0,             // V
                         0.0,             // A
                         0_tps,           // Velocity Limit
                         0_tr_per_s_sq);  // Acceleration Limit

    // TODO: VERIFY THESE CONFIGS
    // Set the second motor to follow the other motor
    m_kickerFollowerMotor.SetControl(
        ctre::phoenix6::controls::Follower(m_kickerMotor.GetDeviceID(), 
                                            ctre::phoenix6::signals::MotorAlignmentValue::Opposed)
    );
}
#pragma endregion

#pragma region SetState
/// @brief Sets the motors to their initial states 
/// @param input The input value for the motors
void Spindexer::SetState(SpindexerState newState)
{
    // Do nothing if the state is the same as the current state
    if (newState == m_state)
        return;

    // Remember the state of the spindexer
    m_state = newState;
    Log("Spindexer", std::string_view{"Setting spindexer state to " + std::to_string(static_cast<int>(newState))});

    // Default the speeds to zero
    auto spindexerSpeed = 0.0_tps;
    auto kickerSpeed    = 0.0_tps;

    // Set the speeds based on the state
    switch (m_state) 
    {
        case SpindexerState::Stopped:
        {
            // Stop both motors (speeds are already default to zero)
            break;
        }

        case SpindexerState::Spindexing:
        {
            // Set both motors to their respective speeds
            spindexerSpeed = SpindexerConstants::SpinnerWheelTurns;
            kickerSpeed    = SpindexerConstants::KickerWheelTurns;
            break;
        }
    }

    Log("Spindexer", std::string_view{"Setting spinner speed to " + std::to_string(spindexerSpeed.value()) + " turns per second and kicker speed to " + std::to_string(kickerSpeed.value()) + " turns per second"});

    // Set the motor speeds
    m_spinnerMotor.SetControl(ctre::phoenix6::controls::VelocityVoltage{spindexerSpeed});
    m_kickerMotor.SetControl(ctre::phoenix6::controls::VelocityVoltage{kickerSpeed});
}
#pragma endregion

File: src/main/java/frc3824/subsystems/Spindexer.java
package frc3824.subsystems;

import com.ctre.phoenix6.controls.VelocityVoltage;
import com.ctre.phoenix6.controls.StrictFollower;
import com.ctre.phoenix6.hardware.TalonFX;

import frc3824.Constants.SpindexerConstants;
import frc3824.Constants.ConstantsCanIds;
import frc3824.lib.TalonFXConfig;

public class Spindexer {

    public enum SpindexerState {
        Stopped,
        Spindexing,
        Backwards
    }

    private final TalonFX m_spinnerMotor        = new TalonFX(ConstantsCanIds.SpinnerMotorId);
    private final TalonFX m_kickerMotor         = new TalonFX(ConstantsCanIds.KickerMotorId);
    private final TalonFX m_kickerFollowerMotor = new TalonFX(ConstantsCanIds.KickerFollowerMotorId);

    private SpindexerState m_state = SpindexerState.Stopped;

    private final VelocityVoltage m_spinnerRequest = new VelocityVoltage(0).withSlot(0);
    private final VelocityVoltage m_kickerRequest  = new VelocityVoltage(0).withSlot(0);

    public Spindexer() {
        TalonFXConfig.configure(m_spinnerMotor,  20.0, true,  false, false, 0.2, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0);
        TalonFXConfig.configure(m_kickerMotor,   20.0, false, false, false, 0.2, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0);
    }

    public void setState(SpindexerState newState) {
        if (newState == m_state) return;

        m_state = newState;

        double spindexerSpeed = 0.0;
        double kickerSpeed    = 0.0;

        switch (m_state) {
            case Stopped:
                break;
            case Spindexing:
                spindexerSpeed = SpindexerConstants.SpinnerWheelTurnsPerSec;
                kickerSpeed    = SpindexerConstants.KickerWheelTurnsPerSec;
                break;
            case Backwards:
                spindexerSpeed = -SpindexerConstants.SpinnerWheelTurnsPerSec;
                kickerSpeed    = -SpindexerConstants.KickerWheelTurnsPerSec;
                break;
        }

        m_spinnerMotor.setControl(m_spinnerRequest.withVelocity(spindexerSpeed));
        m_kickerMotor.setControl(m_kickerRequest.withVelocity(kickerSpeed));
    }
}